lib/source/metadata.js

```diff
@@ -0,0 +1,504 @@
+      pkcs7: data['dynamic/instance-identity/pkcs7'],
+      signature: data['dynamic/instance-identity/signature']
+    };
+
+    // Get instance profile credentials
+    const IAM_METADATA_PATH = 'meta-data/iam/security-credentials/';
+    const credentialsPaths = Object.keys(data).filter((path) =>
+    Path.relative(IAM_METADATA_PATH, path).slice(0, 2) !== '..');
+
+    // Use the first set of credentials
+    if (credentialsPaths.length > 0) {
+      try {
+        const credentials = JSON.parse(data[credentialsPaths[0]]);
+
+        properties.credentials = {
+          lastUpdated: credentials.LastUpdated,
+          type: credentials.Type,
+          accessKeyId: credentials.AccessKeyId,
+          secretAccessKey: credentials.SecretAccessKey,
+          expires: credentials.Expiration
+        };
+      } catch (e) {
+        /* Don't do anything */
+      }
+    }
+
+    // Common meta-data properties
+    ['local-ipv4', 'local-hostname', 'instance-type', 'instance-id', 'hostname',
+      'ami-id', 'public-hostname', 'public-ipv4', 'public-keys', 'reservation-id',
+      'security-groups'].forEach((name) => properties[name] = data[Path.join('meta-data', name)]);
+
+    properties['availability-zone'] = data['meta-data/placement/availability-zone'];
+
+    // Grok the network interface parameters
+    const mac = data['meta-data/mac'];
+
+    if (mac) {
+      const interfacePath = Path.join('meta-data/network/interfaces/macs', mac);
+
+      properties['vpc-id'] = data[Path.join(interfacePath, 'vpc-id')];
+      properties['subnet-id'] = data[Path.join(interfacePath, 'subnet-id')];
+
+      const interfaceProperties = properties.interface = {};
+
+      ['vpc-ipv4-cidr-block', 'subnet-ipv4-cidr-block', 'public-ipv4s', 'mac',
+        'local-ipv4s', 'interface-id'
+      ].forEach((name) => interfaceProperties[name] = data[Path.join(interfacePath, name)]);
+    }
+
+    this.properties = {};
+    this.properties[this.namespace] = properties;
+  }
+}
+
+/**
+ * Metadata Source
+ *
+ * Expose EC2 metadata to the property-set
+ *
+ * @class Source.Metadata
+ * @extends EventEmitter
+ *
+ * @param {Parser} parser
+ * @param {Object} options
+ *      - {Number} interval Polling interval. Default 30s.
+ */
+class Metadata extends EventEmitter {
+  constructor(opts) {
+    super();
+    const options = opts || {};
+
+    if (!options.hasOwnProperty('name')) {
+      options.name = 'ec2-metadata';
+    }
+    if (!options.hasOwnProperty('namespace')) {
+      options.namespace = 'instance';
+    }
+
+    this.interval = DEFAULT_INTERVAL;
+    this._parser = new MetadataParser(options);
+    this.name = options.name || 'source';
+    this._ok = false;
+    this._updated = null;
+
+    /**
+     * Initialize the metadata-service client
+     */
+    this.service = new Aws.MetadataService({
+      httpOptions: {
+        timeout: METADATA_TIMEOUT
+      },
+      host: Config.get('metadata:host')
```

Comment:
You're right. If someone's overwriting the metadata host to force it through a proxy or something they probably know what they're doing.